File: backend/app/utils/url_parser.py
import re
from urllib.parse import urlparse, unquote


def parse_confluence_url(url: str) -> dict:
    """
    Parse a Confluence URL and extract base_url, space_key, and page_id.

    Supports formats:
    - Cloud:  https://company.atlassian.net/wiki/spaces/SPACE/pages/12345/Page+Title
    - Server: https://confluence.company.com/display/SPACE/Page+Title
    - Server: https://confluence.company.com/pages/viewpage.action?pageId=12345
    """
    parsed = urlparse(url)
    base_url = f"{parsed.scheme}://{parsed.netloc}"
    path = unquote(parsed.path)

    result = {
        "base_url": base_url,
        "space_key": None,
        "page_id": None,
        "page_title": None,
    }

    # Cloud format: /wiki/spaces/SPACE/pages/12345/Title
    cloud_match = re.match(
        r"/wiki/spaces/([^/]+)/pages/(\d+)(?:/(.+))?", path
    )
    if cloud_match:
        result["base_url"] = f"{base_url}/wiki"
        result["space_key"] = cloud_match.group(1)
        result["page_id"] = cloud_match.group(2)
        result["page_title"] = cloud_match.group(3)
        return result

    # Data Center / Server format: /spaces/SPACE/pages/12345/Title (no /wiki prefix)
    dc_match = re.match(
        r"/spaces/([^/]+)/pages/(\d+)(?:/(.+))?", path
    )
    if dc_match:
        result["space_key"] = dc_match.group(1)
        result["page_id"] = dc_match.group(2)
        result["page_title"] = dc_match.group(3)
        return result

    # Server format: /display/SPACE/Title
    display_match = re.match(r"/display/([^/]+)/(.+)", path)
    if display_match:
        result["space_key"] = display_match.group(1)
        result["page_title"] = display_match.group(2)
        return result

    # Server format: /pages/viewpage.action?pageId=12345
    if "viewpage.action" in path:
        query_params = dict(
            param.split("=") for param in parsed.query.split("&") if "=" in param
        )
        result["page_id"] = query_params.get("pageId")
        return result

    raise ValueError(f"Could not parse Confluence URL: {url}")

File: electron/src/utils/urlParser.ts
/**
 * URL parser — ported 1:1 from backend/app/utils/url_parser.py.
 *
 * Supports:
 *  - Cloud:  https://company.atlassian.net/wiki/spaces/SPACE/pages/12345/Page+Title
 *  - DC:     https://confluence.company.com/spaces/SPACE/pages/12345/Title
 *  - Server: https://confluence.company.com/display/SPACE/Page+Title
 *  - Server: https://confluence.company.com/pages/viewpage.action?pageId=12345
 */

export interface ParsedConfluenceUrl {
  baseUrl: string;
  spaceKey: string | null;
  pageId: string | null;
  pageTitle: string | null;
}

export function parseConfluenceUrl(rawUrl: string): ParsedConfluenceUrl {
  let parsed: URL;
  try {
    parsed = new URL(rawUrl);
  } catch {
    throw new Error(`Could not parse Confluence URL: ${rawUrl}`);
  }

  const baseUrl = `${parsed.protocol}//${parsed.host}`;
  const path = decodeURIComponent(parsed.pathname);

  const result: ParsedConfluenceUrl = {
    baseUrl,
    spaceKey: null,
    pageId: null,
    pageTitle: null,
  };

  // Cloud: /wiki/spaces/SPACE/pages/12345/Title
  const cloud = /^\/wiki\/spaces\/([^/]+)\/pages\/(\d+)(?:\/(.+))?/.exec(path);
  if (cloud) {
    result.baseUrl = `${baseUrl}/wiki`;
    result.spaceKey = cloud[1];
    result.pageId = cloud[2];
    result.pageTitle = cloud[3] ?? null;
    return result;
  }

  // Data Center: /spaces/SPACE/pages/12345/Title
  const dc = /^\/spaces\/([^/]+)\/pages\/(\d+)(?:\/(.+))?/.exec(path);
  if (dc) {
    result.spaceKey = dc[1];
    result.pageId = dc[2];
    result.pageTitle = dc[3] ?? null;
    return result;
  }

  // Server: /display/SPACE/Title
  const display = /^\/display\/([^/]+)\/(.+)/.exec(path);
  if (display) {
    result.spaceKey = display[1];
    result.pageTitle = display[2];
    return result;
  }

  // Server: /pages/viewpage.action?pageId=12345
  if (path.includes('viewpage.action')) {
    const pageId = parsed.searchParams.get('pageId');
    if (pageId) {
      result.pageId = pageId;
      return result;
    }
  }

  throw new Error(`Could not parse Confluence URL: ${rawUrl}`);
}
